docs/eeps/draft-spi.md

```diff
@@ -0,0 +1,251 @@
+
+Inclusion of `embulk-spi` classes in `embulk-core`
+---------------------------------------------------
+
+The `embulk-core` JAR file includes `embulk-spi` classes extracted in it, not just depends on `embulk-spi`.
+
+This is to prepare for the Java module system in Java 9 and later. The Java module system does not accept the "split packages" situation, in which a Java package and two different JARs each have a class in that package. (See "Migrating Your Library to Java Modules" in [Java Magazine, July/August 2018](https://www.oracle.com/a/ocom/docs/corporate/java-magazine-jul-aug-2018.pdf), pp.53-64, by Nicolai Parlog for handling split packages in the Java module system.)
+
+This means that all `org.embulk.spi` classes must be contained in the same single JAR file. On the other hand, some existing `org.embulk.spi` classes had to move to `embulk-spi` while others had to stay in `embulk-core`. At the same time, they also had to stay in the same Java package for compatibility.
+
+We decided to have all `embulk-spi` classes extracted in the `embulk-core` JAR file so that both of the above requirements are met together. The `embulk-spi` classes are duplicated between the `embulk-spi` JAR file and the `embulk-core` JAR file. In order to avoid conflicts when loading these JAR files, the `pom.xml` of `embulk-core` has a dependency on `embulk-spi` as `provided`.
+
+```xml
+<dependency>
+<groupId>org.embulk</groupId>
+<artifactId>embulk-spi</artifactId>
+<version>0.11</version>
+<scope>provided</scope>
+</dependency>
+```
+
+Selection of SPI classes
+=========================
+
+The following classes and interfaces in `embulk-core` are selected for the Embulk SPI.
+
+1. Interfaces for the main class of each plugin type (`*Plugin` interface such as `InputPlugin` and `ParserPlugin`)
+2. Classes and interfaces that appear in the method signatures of #1, and recursively
+3. Classes and interfaces that need to be shared between plugins
+
+Most of these were simply moved to `embulk-spi`, but some needed special treatment. The rest of this section explains each case.
+
+`org.embulk.spi.Buffer`
+------------------------
+
+The `org.embulk.spi.Buffer` class was an implementation for conveying byte sequences between plugins. We split this into an abstract class `org.embulk.spi.Buffer` in `embulk-spi`, and an implementation class `org.embulk.spi.BufferImpl` in `embulk-core`, so that the Embulk SPI would not have a detailed implementation.
+
+`Buffer` is obviously used by the main classes of file-related plugins, such as File Input plugins and Parser plugins.
+
+The original `Buffer` did not allow direct construction by its constructor. It must be allocated by `BufferAllocator` explained below. Making `Buffer` abstract did not cause any major compatibility problems.
+
+Related changes:
+* [Split `org.embulk.spi.Buffer` into a separate artifact `embulk-api` (embulk/embulk#1232)](https://github.com/embulk/embulk/pull/1232)
```

Comment:
This change Split `org.embulk.spi.Buffer` into a separate artifact **`embulk-api`**. not spi.
We merged `embulk-api` to `embulk-api` What do you think to add #1566 topic?

Reply:
That is mentioned in the later section: https://github.com/embulk/embulk/pull/1657/files/e1aed7ebda36d1aff251df386b9c3bd3523a3b82#diff-b617709ebec63ce9b651a3be990fb5056a74156d29328f3015f8f2afcd38a8f9R235

Reply:
Thank you for your comment!